Announce.php

```diff
@@ -103,4 +105,143 @@ function schema() {
+			function() use ( $t_app, $t_plugin ) {
+				$t_app->post( '/dismiss/{context_id}', [$t_plugin, 'route_dismiss'] );
+				$t_app->get( '/context/{message_id}', [$t_plugin, 'route_new_context'] );
+			}
+		);
+	}
+
+	/**
+	 * RESTful route for Announcement dismissal.
+	 *
+	 * Return status
+	 * - 200 announcement was successfully dismissed
+	 * - 400 invalid context or announcement not dismissable
+	 * - 500 dismissal failed (db insert/update error)
+	 *
+	 * @param Psr\Http\Message\ServerRequestInterface $request
+	 * @param Psr\Http\Message\ResponseInterface $response
+	 * @param array $args
+	 * @return Psr\Http\Message\ResponseInterface
+	 */
+	public function route_dismiss( $request, $response, $args) {
+		$t_timestamp = time();
+		$t_user_id = auth_get_current_user_id();
+
+		# Set the reference Context Id for dismissal
+		$t_context_id = (int)$args['context_id'];
+
+		# Make sure the message context actually exists
+		# and that it can actually be dismissed (i.e. marked as dismissable,
+		# or a time-to-live has been set).
+		$t_context = AnnounceContext::load_by_id( $t_context_id );
+		if( !$t_context ) {
+			return $response->withStatus(
+				HTTP_STATUS_BAD_REQUEST,
+				'Invalid Context Id'
+			);
+		} elseif( !$t_context->dismissable && $t_context->ttl == 0 ) {
+			return $response->withStatus(
+				HTTP_STATUS_BAD_REQUEST,
+				'Non-dismissible announcement'
+			);
+		}
+
+		plugin_push_current( $this->basename );
+		$t_dismissed_table = plugin_table( 'dismissed' );
+
+		# check for existing dismissal
+		$t_query = "SELECT * FROM {$t_dismissed_table}
+			WHERE context_id=" . db_param() . "
+			AND user_id=" . db_param();
+		$t_result = db_query( $t_query,
+			array( $t_context_id, $t_user_id )
+		);
+
+		if( db_num_rows( $t_result ) < 1 ) {
+			$t_query = "INSERT INTO {$t_dismissed_table}
+				(context_id, user_id, timestamp)
+				VALUES
+				(" . db_param() . ", " . db_param() . ", " . db_param(
+				) . ")";
+			$t_param = array( $t_context_id, $t_user_id, $t_timestamp );
+		}
+		else {
+			$t_query = "UPDATE {$t_dismissed_table}
+				SET timestamp = " . db_param() . "
+				WHERE context_id=" . db_param() . " 
+				AND user_id=" . db_param();
+			$t_param = array( $t_timestamp, $t_context_id, $t_user_id );
+		}
+		if( db_query( $t_query, $t_param ) ) {
+			$t_status = HTTP_STATUS_SUCCESS;
+			$t_msg = '';
+		} else {
+			$t_status = HTTP_STATUS_INTERNAL_SERVER_ERROR;
+			$t_msg = 'Announcement dismissal failed';
+		}
+
+		plugin_pop_current();
+
+		return $response->withStatus( $t_status, $t_msg );
+	}
+
+	/**
+	 * RESTful route to generate HTML for new Announcement context.
+	 *
+	 * Return status
+	 * - 200 HTML was successfully generated
+	 * - 400 invalid message
+	 *
+	 * @param Psr\Http\Message\ServerRequestInterface $request
+	 * @param Psr\Http\Message\ResponseInterface $response
+	 * @param array $args
+	 * @return Psr\Http\Message\ResponseInterface
+	 */
+	public function route_new_context( $request, $response, $args) {
```

Comment:
Is there a reason the REST API returns actual HTML rather than json info that is needed to render the HTML form?

Reply:
No particular reason, other than reusing the code from the legacy xmlhttprequest endpoint. Are you suggesting to have the HTML generated in the javascript instead ?

Reply:
Note: the advantage of generating the HTML on the PHP side is that the code can leverage the MantisBT print APIs for option lists. To return raw data in JSON format implies code duplication, so I don't really see the added value.

Reply:
OK, I haven't got to define patterns around returning UI vs. server side rendering.  You can do whatever makes sense for now.  We can always re-visit if it makes sense later.  Just moving to REST API is a good step.